streaming/replication.c

```diff
@@ -0,0 +1,1931 @@
+    info("%s %s [send to %s]: connecting...", REPLICATION_MSG, host->hostname, host->rrdpush_send_destination);
+
+    rep_state->socket = connect_to_one_of(
+            host->rrdpush_send_destination
+            , default_port
+            , &tv
+            , &rep_state->reconnects_counter
+            , rep_state->connected_to
+            , sizeof(rep_state->connected_to)-1
+    );
+
+    if(unlikely(rep_state->socket == -1)) {
+        error("%s %s [send to %s]: failed to connect", REPLICATION_MSG, host->hostname, host->rrdpush_send_destination);
+        return 0;
+    }
+
+    info("%s %s [send to %s]: initializing communication...", REPLICATION_MSG, host->hostname, rep_state->connected_to);
+
+#ifdef ENABLE_HTTPS
+    if( netdata_replication_client_ctx ){
+        rep_state->ssl.flags = NETDATA_SSL_START;
+        if (!rep_state->ssl.conn){
+            rep_state->ssl.conn = SSL_new(netdata_replication_client_ctx);
+            if(!rep_state->ssl.conn){
+                error("Failed to allocate SSL structure.");
+                rep_state->ssl.flags = NETDATA_SSL_NO_HANDSHAKE;
+            }
+        }
+        else{
+            SSL_clear(rep_state->ssl.conn);
+        }
+
+        if (rep_state->ssl.conn)
+        {
+            if (SSL_set_fd(rep_state->ssl.conn, rep_state->socket) != 1) {
+                error("Failed to set the socket to the SSL on socket fd %d.", rep_state->socket);
+                rep_state->ssl.flags = NETDATA_SSL_NO_HANDSHAKE;
+            } else{
+                rep_state->ssl.flags = NETDATA_SSL_HANDSHAKE_COMPLETE;
+            }
+        }
+    }
+    else {
+        rep_state->ssl.flags = NETDATA_SSL_NO_HANDSHAKE;
+    }
+#endif
+
+    stream_encoded_t se;
+    rrdpush_encode_variable(&se, host);
+    char http[HTTP_HEADER_SIZE + 1];
+    int eol = snprintfz(http, HTTP_HEADER_SIZE,
+            "%s "
+            "key=%s"
+            "&hostname=%s"
+            "&registry_hostname=%s"
+            "&machine_guid=%s"
+            "&update_every=%d"
+            "&timezone=%s"
+            "&abbrev_timezone=%s"
+            "&utc_offset=%d"
+            "&hops=%d"
+            "&tags=%s"
+            "&ver=%u"
+            "&NETDATA_PROTOCOL_VERSION=%s"
+            " HTTP/1.1\r\n"
+            "User-Agent: %s/%s\r\n"
+            "Accept: */*\r\n\r\n"
+            , REPLICATE_CMD
+            , host->rrdpush_send_api_key
+            , host->hostname
+            , host->registry_hostname
+            , host->machine_guid
+            , default_rrd_update_every
+            , host->timezone
+            , host->abbrev_timezone
+            , host->utc_offset
+            , host->system_info->hops + 1
+            , (host->tags) ? host->tags : ""
+            , STREAMING_PROTOCOL_CURRENT_VERSION
+            , STREAMING_PROTOCOL_VERSION
+            , host->program_name
+            , host->program_version);
+    http[eol] = 0x00;
```

Comment:
The function `snprintfz` will set the `NULL` value, so this line is not necessary.